Add a stream serializer

Matthew-Whitlock opened the pull request:
Enables serializing to/from an open stream.

Matthew-Whitlock committed:
Add stream_serializer

Matthew-Whitlock committed:
Update documentation

PhilMiller commented on src/checkpoint/serializers/stream_serializer.h:
```diff
@@ -0,0 +1,111 @@
+//@HEADER
+// *****************************************************************************
+//
+//                             memory_serializer.h
```
Header name needs to be corrected here.

PhilMiller commented:
Please include some suitable unit tests covering the new serializer

The code all looks fine to me

Matthew-Whitlock committed:
Add stream serializer tests

Also removed deprecated Kokkos header includes.

Matthew-Whitlock committed:
Fix header comment

nmm0 commented on src/checkpoint/serializers/stream_serializer.h:
```diff
@@ -0,0 +1,111 @@
+
+template <typename StreamT, eSerializationMode mode>
+struct StreamSerializer : Serializer {
+  //This really doesn't expect to be initialized like this, disabling until someone yells at me.
```
You can remove dead code

nmm0 reviewed (commented):
Guard your code with `#if __cplusplus > 201703L` to compile only for C++17

Matthew-Whitlock committed:
Replace constexpr uses

Matthew-Whitlock committed:
Merge branch 'develop' into stream_serializer

Matthew-Whitlock committed:
Merge branch 'develop' into stream_serializer

Conflicts:
	src/checkpoint/container/view_equality.h
	tests/unit/test_commons.h

Matthew-Whitlock committed:
Add virtual support for StreamSerializer

Matthew-Whitlock committed:
Add includes for basic ostream/istream, for virtual serializer default templates

Matthew-Whitlock committed:
Fixes for developing with spack

Explicitly link to libpthread etc in the cmake-ian way
Also gitignore the files it places in source (why would they do this to me?)

Matthew-Whitlock committed:
Fix syntax issues with clang compiler

Matthew-Whitlock committed:
Fix problem when running parallel tests, tests use different file names

Matthew-Whitlock commented:
Replaced with #329
